### PR TITLE
Use evidence period when pruning

### PR DESCRIPTION
**Pull request by samricotta**
Resolves issue #9391 

Implements logic that uses the evidence period to indicate whether or not to prune unnecessary data such as abci commits, consensus params, validator sets. This way we can successfully check we have the required data to prove malicious behaviour.

### CODE REVIEW

File: store/store.go [hunk 1]
```diff
@@ -300,26 +302,41 @@ func (bs *BlockStore) PruneBlocks(height int64) (uint64, error) {
...
 		}
-		if err := batch.Delete(calcBlockMetaKey(h)); err != nil {
-			return 0, err
+
+		// This logic is in place to protect data that proves malicious behavior.
+		// Instead of pruning all blocks we only prune the meta and commit data
+		// if the evidence has exceeded the block height.
+
+		if evidencePoint == height && !evidence.IsEvidenceExpired(state.LastBlockHeight, state.LastBlockTime, h, meta.Header.Time, state.ConsensusParams.Evidence) {
+			evidencePoint = h
```
Comment (jmalicevic) [inline store/store.go:318]: Could you not just break here? Once you discover the evidence point you do not delete anything except flush the existing batch? And line 326  `batch.Delete(calcBlockHashKey(meta.BlockID.Hash))` should not execute at all, same for line 334 (`err := batch.Delete(calcSeenCommitKey(h))`)? For the code on these lines tehre may be context I am missing so maybe I am wrong here.

Comment (cmwaters) [reply store/store.go:318]: Once we pass the evidence point, we are still pruning the entire block and the seen commit

Comment (cmwaters) [reply store/store.go:318]: This data is not needed for evidence verification

File: store/store.go [hunk 2]
```diff
@@ -300,26 +302,41 @@ func (bs *BlockStore) PruneBlocks(height int64) (uint64, error) {
...
+
 		meta := bs.LoadBlockMeta(h)
 		if meta == nil { // assume already deleted
 			continue
 		}
-		if err := batch.Delete(calcBlockMetaKey(h)); err != nil {
-			return 0, err
+
+		// This logic is in place to protect data that proves malicious behavior.
+		// Instead of pruning all blocks we only prune the meta and commit data
```
Comment (cmwaters) [inline store/store.go:314]:
```suggestion
		// If the height is within the evidence age, we continue to persist the header and commit data
```

File: store/store.go [hunk 3]
```diff
@@ -300,26 +302,41 @@ func (bs *BlockStore) PruneBlocks(height int64) (uint64, error) {
...
+		// if the evidence has exceeded the block height.
```
Comment (cmwaters) [inline store/store.go:315]:
```suggestion
		// If the height is within the evidence age, we continue to persist the header and commit data
```

File: state/store_test.go [hunk 4]
```diff
@@ -180,6 +182,10 @@ func TestPruneStates(t *testing.T) {
 					require.NoError(t, err, "validators height %v", h)
 					require.NotNil(t, vals)
 				} else {
+					fmt.Println(sliceToMap(tc.expectVals))
+					fmt.Println(h)
+					fmt.Println(vals)
+					fmt.Println(expectABCI)
```
Comment (cmwaters) [inline state/store_test.go:188]: This was just for testing right?

File: evidence/verify.go [hunk 5]
```diff
@@ -34,10 +33,10 @@ func (evpool *Pool) verify(evidence types.Evidence) error {
 		return fmt.Errorf("evidence has a different time to the block it is associated with (%v != %v)",
 			evidence.Time(), evTime)
 	}
-	ageDuration := state.LastBlockTime.Sub(evTime)
 
-	// check that the evidence hasn't expired
-	if ageDuration > evidenceParams.MaxAgeDuration && ageNumBlocks > evidenceParams.MaxAgeNumBlocks {
+	// checking if evidence is expired calculated using the block evidence time and height
+
```
Comment (cmwaters) [inline evidence/verify.go:38]:
```suggestion
```
nit: remove blank line

File: state/store.go [hunk 6]
```diff
@@ -71,7 +71,7 @@ type Store interface {
 	// Bootstrap is used for bootstrapping state when not starting from a initial height.
 	Bootstrap(State) error
 	// PruneStates takes the height from which to start prning and which height stop at
```
Comment (cmwaters) [inline state/store.go:73]:
```suggestion
	// PruneStates takes the height from which to start pruning and which height stop at
	// and removes all state-related data. It also takes the height where evidence may need to
	// be verified to ensure that the validator sets at those heights are not pruned
```

File: state/store_test.go (hunk 4 above)
Comment (samricotta) [reply state/store_test.go:188]: lol

File: store/store.go (hunk 1 above)
Comment (samricotta) [reply store/store.go:318]: This may have been confusing with my comments coupled wit the logic implemented

Comment (samricotta) [reply store/store.go:318]: have added some more comments which might shed some light

File: state/store.go [hunk 7]
```diff
@@ -237,14 +237,15 @@ func (store dbStore) Bootstrap(state State) error {
...
-func (store dbStore) PruneStates(from int64, to int64) error {
+func (store dbStore) PruneStates(from int64, to int64, evidenceThresholdHeight int64) error {
 	if from <= 0 || to <= 0 {
 		return fmt.Errorf("from height %v and to height %v must be greater than 0", from, to)
 	}
 	if from >= to {
 		return fmt.Errorf("from height %v must be lower than to height %v", from, to)
 	}
-	valInfo, err := loadValidatorsInfo(store.db, to)
+
```
Comment (cmwaters) [inline state/store.go:247]:
```suggestion
    // we load what will be the validator set of the lowest height (after pruning) so we can ensure that we 
    // have the necessary data to retrieve it in the future
```

File: state/store_test.go [hunk 8]
```diff
@@ -88,23 +88,25 @@ func BenchmarkLoadValidators(b *testing.B) {
...
+		"error when from > to":         {100, 3, 2, 2, true, nil, nil, nil},
+		"error when from == to":        {100, 3, 3, 3, true, nil, nil, nil},
+		"error when to does not exist": {100, 1, 101, 101, true, nil, nil, nil},
+		"prune all":                    {100, 1, 100, 100, false, []int64{93, 100}, []int64{95, 100}, []int64{100}},
+		"prune some": {10, 2, 8, 8, false, []int64{1, 3, 8, 9, 10},
 			[]int64{1, 5, 8, 9, 10}, []int64{1, 8, 9, 10}},
-		"prune across checkpoint": {100001, 1, 100001, false, []int64{99993, 100000, 100001},
+		"prune across checkpoint": {100001, 1, 100001, 100001, false, []int64{99993, 100000, 100001},
 			[]int64{99995, 100001}, []int64{100001}},
+		"prune when evidence height < height": {20, 1, 18, 17, false, []int64{13, 17, 18, 19, 20}, []int64{15, 18, 19, 20}, []int64{18, 19, 20}},
```
Comment (cmwaters) [inline state/store_test.go:109]: can we also add a case where evidence height is less than the "from" height

**Review by cmwaters — APPROVED** [review]
LGTM apart from a few nits

File: store/store.go (hunk 1 above)
Comment (jmalicevic) [reply store/store.go:317]: Thanks for the clarification, makes sense.

**Pull request merged**